### PR TITLE
STORM-3602 fix switching on low water mark for loadaware shuffle

### PR DESCRIPTION
**Pull request by agresch**

### CODE REVIEW

File: storm-client/src/jvm/org/apache/storm/grouping/LoadAwareShuffleGrouping.java [hunk 1]
```diff
@@ -305,7 +309,7 @@ public static Scope downgrade(Scope current) {
             }
         }
 
-        public static Scope upgrade(Scope current) {
+        public static LocalityScope upgradeFromWorker(LocalityScope current) {
```
Comment (Ethanlm) [inline storm-client/src/jvm/org/apache/storm/grouping/LoadAwareShuffleGrouping.java]: This name can be a little confusing.  `upgradeFromWorker` implies the current scope is `WORKER_LOCAL`.

I would prefer to keep 'downgrade` and `upgrade`.

File: storm-client/src/jvm/org/apache/storm/grouping/LoadAwareShuffleGrouping.java [hunk 2]
```diff
@@ -47,13 +47,14 @@
     private Random random;
     private volatile int[] prepareChoices;
     private AtomicInteger current;
-    private Scope currentScope;
+    @VisibleForTesting
+    LocalityScope currentScope;
```
Comment (Ethanlm) [inline storm-client/src/jvm/org/apache/storm/grouping/LoadAwareShuffleGrouping.java]: I would keep this as private and add a `getCurrentScope` function for test only with `@VisibleForTesting` annotation.

File: storm-client/test/jvm/org/apache/storm/grouping/LoadAwareShuffleGroupingTest.java [hunk 3]
```diff
@@ -496,4 +499,75 @@ public Long call() throws Exception {
 
         refreshService.shutdownNow();
     }
+
+    @Test
+    public void testLoadSwitching() throws Exception {
```
Comment (Ethanlm) [inline storm-client/test/jvm/org/apache/storm/grouping/LoadAwareShuffleGroupingTest.java]: Good unit test!

**Comment by agresch** [comment]
@Ethanlm - requested changes made

**Pull request merged**